pkg/executor/sortexec/sort_spill.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sortexec
+
+import (
+	"sync"
+
+	"github.com/pingcap/tidb/pkg/util/chunk"
+	"github.com/pingcap/tidb/pkg/util/logutil"
+	"github.com/pingcap/tidb/pkg/util/memory"
+	"go.uber.org/zap"
+)
+
+type spillHelper struct {
+	syncLock sync.Mutex
+
+	lock            *sync.Mutex
+	cond            sync.Cond
+	spillError      error
+	isSpilling      bool
+	isSpillTrigered bool
+}
+
+func newSpillHelper() *spillHelper {
+	lock := new(sync.Mutex)
+	return &spillHelper{
+		lock:            lock,
+		cond:            *sync.NewCond(lock),
+		spillError:      nil,
+		isSpilling:      false,
+		isSpillTrigered: false,
+	}
+}
+
+func (s *spillHelper) reset() {
+	s.isSpilling = false
+	s.isSpillTrigered = false
+}
+
+func (s *spillHelper) setIsSpilling(isSpilling bool) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if isSpilling {
+		s.setSpillTriggeredNoLock()
+	}
+	s.isSpilling = isSpilling
+}
+
+func (s *spillHelper) setSpillTriggeredNoLock() {
+	s.isSpillTrigered = true
+}
+
+func (s *spillHelper) isSpillTriggeredNoLock() bool {
+	return s.isSpillTrigered
+}
```

Comment:
how about using `atomic_bool`

Reply:
> how about using `atomic_bool`

I think it's needless as it has been protected with other fields by the lock together.